### PR TITLE
Feature/previous exam accommodations

### PR DESCRIPTION
**Pull request by millerts**
This is the code that initializes the exam accommodations when opening a previous exam.

### CODE REVIEW

File: client/src/main/java/tds/exam/ExamAccommodation.java [hunk 1]
```diff
@@ -12,12 +12,17 @@
...
     private String type;
     private String code;
+    private String value;
     private String description;
+    private boolean selectable;
+    private boolean allowChange;
     private Instant deniedAt;
     private Instant createdAt;
     private Instant deletedAt;
+    private boolean multipleToolTypes;
```
Comment (jtreuting) [inline client/src/main/java/tds/exam/ExamAccommodation.java]: Is there a benefit to storing this as a boolean, which is specific to the use case as I remember, instead of `int numToolTypes` with a helper getter like `boolean isMultipleToolTypes() { return numToolTypes > 1 }`?

File: service/src/main/java/tds/exam/repositories/impl/ExamQueryRepositoryImpl.java [hunk 2]
```diff
@@ -67,7 +68,8 @@ public ExamQueryRepositoryImpl(@Qualifier("queryJdbcTemplate") NamedParameterJdb
             "   ee.abnormal_starts, \n" +
             "   ee.waiting_for_segment_approval, \n" +
             "   ee.current_segment_position, \n" +
-            "   e.created_at, \n" +
+            "   ee.custom_accommodations, \n" +
+            "   ee.language, \n" +
```
Comment (jtreuting) [inline service/src/main/java/tds/exam/repositories/impl/ExamQueryRepositoryImpl.java]: Same question as above, is this language or languageCode?

File: service/src/main/java/tds/exam/services/impl/ExamAccommodationServiceImpl.java [hunk 3]
```diff
@@ -70,4 +84,153 @@ public ExamAccommodationServiceImpl(ExamAccommodationQueryRepository examAccommo
...
+         */
+        List<ExamAccommodation> examAccommodations = findAllAccommodations(exam.getId());
+        if (examAccommodations.isEmpty()) {
+            initializeExamAccommodations(exam);
+        } else {
+            initializePreviousAccommodations(exam, assessment, segmentPosition, restoreRts, guestAccommodations, examAccommodations);
+        }
+
+        //Fetch the updated exam accommodations
+        examAccommodations = findAllAccommodations(exam.getId());
```
Comment (jtreuting) [inline service/src/main/java/tds/exam/services/impl/ExamAccommodationServiceImpl.java]: Is it possible to return the accommodations from the initialize methods called right before this to save a DB trip?

File: service/src/main/java/tds/exam/services/impl/ExamAccommodationServiceImpl.java [hunk 4]
```diff
@@ -70,4 +84,153 @@ public ExamAccommodationServiceImpl(ExamAccommodationQueryRepository examAccommo
...
+
+        //StudentDLL lines 6967 - 6875
```
Comment (jtreuting) [inline service/src/main/java/tds/exam/services/impl/ExamAccommodationServiceImpl.java]: On develop, I'm seeing the line numbers as either 6857 - 6868 or maybe 6865 - 6868

File: service/src/main/java/tds/exam/services/impl/ExamAccommodationServiceImpl.java [hunk 5]
```diff
@@ -70,4 +84,153 @@ public ExamAccommodationServiceImpl(ExamAccommodationQueryRepository examAccommo
...
+            .toArray(ExamAccommodation[]::new);
+
+        examAccommodationCommandRepository.update(examAccommodationsToDenyApproval);
+    }
+
+    private static String getOtherAccommodationValue(String formattedValue) {
+        return formattedValue.substring("TDS_Other#".length());
+    }
+
+    private List<String> splitAccommodationCodes(String accommodationFamily, String guestAccommodations) {
```
Comment (jtreuting) [inline service/src/main/java/tds/exam/services/impl/ExamAccommodationServiceImpl.java]: Let's run thru this when we do our review meeting tomorrow.  The `_SplitAccomCodes_FN` method is hard to follow.

File: service/src/main/java/tds/exam/services/impl/ExamAccommodationServiceImpl.java [hunk 6]
```diff
@@ -70,4 +84,153 @@ public ExamAccommodationServiceImpl(ExamAccommodationQueryRepository examAccommo
...
+         This replaces the functionality of the following bits of code
+         - StudentDLL 6834 - 6843
+         - StudentDLL _InitOpportunityAccommodations_SP
+         - CommonDLL _UpdateOpportunityAccommodations_SP
+         */
+        List<ExamAccommodation> examAccommodations = findAllAccommodations(exam.getId());
+        if (examAccommodations.isEmpty()) {
+            initializeExamAccommodations(exam);
+        } else {
+            initializePreviousAccommodations(exam, assessment, segmentPosition, restoreRts, guestAccommodations, examAccommodations);
```
Comment (jtreuting) [inline service/src/main/java/tds/exam/services/impl/ExamAccommodationServiceImpl.java]: What is `segmenetPosition` here?  I see in `initializePreviousAccommodations` it matches on segment (which is 0 for the entire assessment and X for a specific segment, as I recall).  I think I'm confused what would be passed in here.  Isn't this initializing accommodations for an exam based on a previous exam?  So would it need to do it for all segments?

File: service/src/main/java/tds/exam/services/impl/ExamAccommodationServiceImpl.java [hunk 7]
```diff
@@ -70,4 +84,153 @@ public ExamAccommodationServiceImpl(ExamAccommodationQueryRepository examAccommo
...
+                    .withAllowChange(false)
+                    .withSelectable(false)
+                    .withSegmentPosition(segmentPosition)
+                    .build();
+
+                break;
+            }
+        }
+
+        final Optional<ExamAccommodation> maybeOtherExamAccommodation = Optional.ofNullable(otherExamAccommodation);
```
Comment (jtreuting) [inline service/src/main/java/tds/exam/services/impl/ExamAccommodationServiceImpl.java]: It feels like the use of Optional here is not needed since it is within the internals of the method and not used to convey meaning as the return type of the service.  Using `otherExamAccommodation` below seems just as clean to me:  `.filter(accommodation -> otherExamAccommodation == null || accommodation.getType().equals(otherExamAccommodation.getType()))`

File: client/src/main/java/tds/exam/Exam.java [hunk 8]
```diff
@@ -64,7 +66,8 @@
         private int abnormalStarts;
         private boolean waitingForSegmentApproval;
         private int currentSegmentPosition;
-
+        private String language;
```
Comment (millerts) [inline client/src/main/java/tds/exam/Exam.java]: Yes, this should probably be language code.  I used language because that is what legacy stores.  I can make the change.

Comment (burzum619) [inline client/src/main/java/tds/exam/Exam.java]: I believe it is "languageCode" technically, which I've added in my PR for initializeSegments

File: client/src/main/java/tds/exam/ExamAccommodation.java (hunk 1 above)
Comment (millerts) [inline client/src/main/java/tds/exam/ExamAccommodation.java]: I'll change it to store the int.  I think this check is only done once so a convenience method isn't needed.

File: service/src/main/java/tds/exam/services/impl/ExamAccommodationServiceImpl.java (hunk 6 above)
Comment (millerts) [inline service/src/main/java/tds/exam/services/impl/ExamAccommodationServiceImpl.java]: It is 0 in the legacy code. I didn't sweep through the legacy code for all the calls to the method this is replacing.  I can do so.

File: client/src/main/java/tds/exam/Exam.java (hunk 8 above)
Comment (millerts) [inline client/src/main/java/tds/exam/Exam.java]: In the legacy code it is the language not the code.  I left it as is because I wanted to avoid all the churn that was happening with language and assessment.  Now that that work is complete I will make the necessary change to use language code.
